consensus/src/threshold_simplex/signing_scheme/mod.rs

```diff
@@ -0,0 +1,179 @@
+/// available, derives a randomness seed for leader rotation. Implementations may override
+/// the provided defaults to take advantage of scheme-specific batching strategies.
+pub trait SigningScheme: Clone + Debug + Send + Sync + 'static {
+    /// Vote signature emitted by individual validators.
+    type Signature: Clone + Debug + PartialEq + Eq + Hash + Send + Sync + CodecFixed<Cfg = ()>;
+    /// Quorum certificate recovered from a set of votes.
+    type Certificate: Clone + Debug + PartialEq + Eq + Hash + Send + Sync + Codec;
+    /// Randomness seed derived from a certificate, if the scheme supports it.
+    type Seed: Clone + Encode + Send;
+
+    /// Converts the scheme into a pure verifier.
+    ///
+    /// The returned instance should return `false` from `can_sign()`.
+    fn into_verifier(self) -> Self;
+
+    /// Returns `true` if this instance holds the secrets required to author votes.
+    fn can_sign(&self) -> bool;
+
+    /// Signs a vote for the given context using the supplied namespace for domain separation.
+    fn sign_vote<D: Digest>(&self, namespace: &[u8], context: VoteContext<'_, D>) -> Vote<Self>;
+
+    /// Verifies a single vote against the participant material managed by the scheme.
+    fn verify_vote<D: Digest>(
+        &self,
+        namespace: &[u8],
+        context: VoteContext<'_, D>,
+        vote: &Vote<Self>,
+    ) -> bool;
+
+    /// Batch-verifies votes and separates valid messages from the voter indices that failed
+    /// verification.
+    ///
+    /// Callers must not include duplicate votes from the same signer.
+    fn verify_votes<R, D, I>(
+        &self,
+        _rng: &mut R,
+        namespace: &[u8],
+        context: VoteContext<'_, D>,
+        votes: I,
+    ) -> VoteVerification<Self>
+    where
+        R: Rng + CryptoRng,
+        D: Digest,
+        I: IntoIterator<Item = Vote<Self>>,
+    {
+        let mut invalid = BTreeSet::new();
+
+        let verified = votes.into_iter().filter_map(|vote| {
+            if self.verify_vote(namespace, context, &vote) {
+                Some(vote)
+            } else {
+                invalid.insert(vote.signer);
+                None
+            }
+        });
+
+        VoteVerification::new(verified.collect(), invalid.into_iter().collect())
+    }
+
+    /// Aggregates a quorum of votes into a certificate, returning `None` if the quorum is not met.
+    ///
+    /// `certificate` carries a previously recovered certificate for the same proposal (in
+    /// a different context), when available. Schemes such as threshold BLS can reuse part
+    /// of that certificate (e.g. the seed signature from a notarization) when assembling
+    /// a finalization certificate, while most other schemes may simply ignore it.
+    ///
+    /// Callers must not include duplicate votes from the same signer.
+    fn assemble_certificate<I>(
+        &self,
+        votes: I,
+        certificate: Option<Self::Certificate>,
```

Comment:
nit: tempted to suggest we drop this optimization for now. I think its the ~only confusing part of this refactor.

Reply:
Removed it.